clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleZonalServerGroupCachingAgent.java

```diff
@@ -0,0 +1,906 @@
+        && instanceTemplate.getProperties().getMetadata() != null
+        && instanceTemplate.getProperties().getMetadata().getItems() != null) {
+
+      ImmutableMap<String, String> metadata =
+          instanceTemplate.getProperties().getMetadata().getItems().stream()
+              .filter(item -> item.getKey() != null)
+              .filter(item -> item.getValue() != null)
+              .collect(toImmutableMap(Items::getKey, Items::getValue));
+
+      if (metadata.containsKey(GLOBAL_LOAD_BALANCER_NAMES)) {
+        autoscalerGroup.put(
+            GLOBAL_LOAD_BALANCER_NAMES,
+            COMMA.splitToList(metadata.get(GLOBAL_LOAD_BALANCER_NAMES)));
+      }
+
+      if (metadata.containsKey(REGIONAL_LOAD_BALANCER_NAMES)) {
+        autoscalerGroup.put(
+            REGIONAL_LOAD_BALANCER_NAMES,
+            COMMA.splitToList(metadata.get(REGIONAL_LOAD_BALANCER_NAMES)));
+        List<String> loadBalancerNames =
+            Utils.deriveNetworkLoadBalancerNamesFromTargetPoolUrls(manager.getTargetPools());
+
+        // The isDisabled property of a server group is set based on whether there are associated
+        // target pools,
+        // and whether the metadata of the server group contains a list of load balancers to
+        // actually
+        // associate
+        // the server group with.
+        // We set the disabled state for L4 lBs here (before writing into the cache) and calculate
+        // the L7 disabled state when we read the server groups from the cache.
+        serverGroup.setDisabled(loadBalancerNames.isEmpty());
+      }
+
+      if (metadata.containsKey(BACKEND_SERVICE_NAMES)) {
+        autoscalerGroup.put(
+            BACKEND_SERVICE_NAMES, COMMA.splitToList(metadata.get(BACKEND_SERVICE_NAMES)));
+      }
+
+      if (metadata.containsKey(LOAD_BALANCING_POLICY)) {
+        try {
+          autoscalerGroup.put(
+              LOAD_BALANCING_POLICY,
+              objectMapper.readValue(
+                  metadata.get(LOAD_BALANCING_POLICY), GoogleHttpLoadBalancingPolicy.class));
+        } catch (IOException e) {
+          log.warn("Error parsing load balancing policy", e);
+        }
+      }
+    }
+
+    serverGroup.setAsg(autoscalerGroup);
+  }
+
+  private static void populateAutoscaler(
+      GoogleServerGroup serverGroup, @Nullable Autoscaler autoscaler) {
+
+    if (autoscaler == null) {
+      return;
+    }
+
+    AutoscalingPolicy autoscalingPolicy = autoscaler.getAutoscalingPolicy();
+    if (autoscalingPolicy != null) {
+      serverGroup.setAutoscalingPolicy(autoscalingPolicy);
+      serverGroup.getAsg().put("minSize", autoscalingPolicy.getMinNumReplicas());
+      serverGroup.getAsg().put("maxSize", autoscalingPolicy.getMaxNumReplicas());
+    }
+    if (autoscaler.getStatusDetails() != null) {
+      serverGroup.setAutoscalingMessages(
+          autoscaler.getStatusDetails().stream()
+              .filter(details -> details.getMessage() != null)
```

Comment:
nit: I might invert the `filter` and `map` since we're just filtering on the result of the `map`

Reply:
Good call, looks much nicer too:
```
          autoscaler.getStatusDetails().stream()
              .map(AutoscalerStatusDetails::getMessage)
              .filter(Objects::nonNull)
              .collect(toImmutableList())
```